clash-cores/src/Clash/Cores/Xilinx/DcFifo/BlackBoxes/Internal.hs

```diff
@@ -0,0 +1,217 @@
+
+import Prelude
+
+import Clash.Core.Literal (Literal(NaturalLiteral))
+import Clash.Core.TermLiteral (termToDataError, TermLiteralSNat(..))
+import Clash.Core.Term (Term(Literal))
+import qualified Clash.Primitives.DSL as DSL
+import Clash.Netlist.BlackBox.Types (BlackBoxFunction, emptyBlackBoxMeta)
+import Clash.Netlist.Types (TemplateFunction(..), BlackBox(BBFunction))
+import Clash.Netlist.Util (orNothing)
+
+import qualified Clash.Netlist.Id as Id
+import qualified Clash.Netlist.Types as N
+import qualified Clash.Netlist.BlackBox.Types as N
+
+import Control.Monad (unless)
+import Data.Either (lefts)
+import Data.Maybe (catMaybes)
+import GHC.Stack (HasCallStack)
+
+import Clash.Cores.Xilinx.DcFifo.Implementation (DcConfig(..))
+import Clash.Cores.Xilinx.DcFifo.Instances ()
+import Clash.Cores.Xilinx.Internal (toTclBool, renderTcl, defIpConfig, IpConfig (properties))
+
+-- | Blackbox function for 'Clash.Cores.Xilinx.Fifo.dcFifo'. It parses the "DcConfig"
+-- supplied to 'dcFifo' from its Term representation, and passes them to two
+-- template functions:
+--
+--   * 'dcFifoTclTF': renders TCL file calling Xilinx's `create_ip`
+--   * 'dcFifoTF': instantiates IP generated in 'dcFifoTclTF'
+--
+-- Additionally, it generates a unique module name for the Xilinx IP.
+--
+dcFifoBBF :: HasCallStack => BlackBoxFunction
+dcFifoBBF _isD _primName args _resTys = do
+  let
+    [  _knownDomainWrite, _knownDomainRead, knownNatN
+     , _bitPack, _, _nfDataX, _knownNatDepth
+     , _constraint1, _constraint2
+     , either error id . termToDataError -> dcConfig
+     , _wClk, _rClk, _rst, _wData
+     , _rEnable
+
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    --  , _, _, _, _, _, _, _
+     ] = lefts args
+
+  pure $ case knownNatN of
+    Literal (NaturalLiteral n) ->
+      let width = fromInteger n in
+      Right (bbMeta width dcConfig, bb dcConfig)
+    _ -> Left "Unexpected type of knownNatN"
+
+ where
+  bbMeta width dcConfig = emptyBlackBoxMeta
+    { N.bbKind = N.TDecl
+    , N.bbIncludes =
+      [ ( ("dcfifo", "tcl")
+        , BBFunction (show 'dcFifoTclTF) 0 (dcFifoTclTF width dcConfig))
+      ]
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    -- , N.bbResultNames =
+    --   , N.BBTemplate [N.Text "wr_full"]
+    --   , N.BBTemplate [N.Text "wr_data_count"]
+
+    --   , N.BBTemplate [N.Text "rd_empty"]
+    --   , N.BBTemplate [N.Text "rd_data_count"]
+    --   , N.BBTemplate [N.Text "rd_dout"]
+    --   ]
+    }
+
+  bb :: DcConfig TermLiteralSNat -> BlackBox
+  bb dcConfig = BBFunction (show 'dcFifoTF) 0 (dcFifoTF dcConfig)
+
+-- | Instantiate IP generated with 'dcFifoTclTF'
+dcFifoTF :: DcConfig TermLiteralSNat -> TemplateFunction
+dcFifoTF DcConfig{..} = TemplateFunction [0..6] (const True) $ \bbCtx -> do
```

Comment:
`[0..6]` doesn't make sense, you've got two lists of arguments you use:

```
   [  _knownDomainWrite, _knownDomainRead, knownNatN
     , _bitPack, _, _nfDataX, _knownNatDepth
     , _constraint1, _constraint2
     , either error id . termToDataError -> dcConfig
     , _wClk, _rClk, _rst, _wData
     , _rEnable

     , _, _, _, _, _, _, _
     ] = lefts args
```

So their indices would  be:

```
   [  _0, _1, 2
     , _3, _4, _5, _6
     , _7, _8
     , dcConfig9
     , _10, _11, _12, _13
     , _14

     , _, _, _, _, _, _, _
     ] = lefts args
```

just looking at this list, you use `[2, 9]`. If you take the other list into account, you get `[2, 9, 10, 11, 12, 13, 14]` or:

* 2: knownNatN argument
* 9: dcFifo argument
* 10: write clock
* 11: read clock
* 12: reset
* 13: write data
* 14: read enable

does that make sense? :)